Handle subpixel viewboxes for SVG skins

adroitwhiz opened the pull request:
## Depends on #745

### Resolves

Resolves #143
Resolves #233
Resolves #427
Resolves #533
Resolves #639

### Proposed Changes

- ~~Clean up the SVGSkin code a bit~~ Done in #649
- Set the minimum SVG texture scale to 1/8 its normal size
  - This is mentioned in #550--the "native size" of the rendered sprite must be a multiple of 1/the smallest scale, so this means that all SVG skins' dimensions will be rounded up to a multiple of 8. Setting this lower (to, for instance, 1/32 scale) would require all SVG skins' dimensions to be rounded up to a multiple of 32. I think 1/8 is a good tradeoff between skins not looking pixelated at small sizes and minimizing the amount of "slack space" around the edges.
- Use the adjusted bounds + center from the algorithm in #550 on SVG skins.
- Pass "logical bounds" into the shader so it can calculate "logical coordinates" under which distortion effects are applied.
  - SVG skins' bounds are now expanded, but we want things like the whirl and fisheye effects' center to still be positioned in the middle of the original, non-expanded bounds. We also don't want the pixelate or mosaic effects to go beyond the non-expanded bounds. To do this, we tell the shader what the original bounds are.
- Separate skins' size and rotation center into a "native size" + "native rotation center" and "quad size" + "quad rotation center'.
  - The "native size" and "native rotation center" refer to the un-expanded bounds. They are used to calculate skins' AABBs and when the VM wants to get a skin's size. They do not depend on details like the amount of "slack space" present around the adjusted texture, and as such are used for fencing sprites.
  - The "quad size" and "quad rotation center" refer to the expanded bounds. They are used for everything else, and are so named because they refer to the size of the actual rendered quadrilateral. I'm not sure what else to call them-- "texture size" seems misleading. Any ideas?

### Reason for Changes

As I talked about in #550, these changes seem to be the best way to fix the jittering problem that vector costumes have at the moment. These do add some complexity but I'm not sure how to avoid it, so I could use your help in that area.

### Test Coverage

Tests ~~are blocked on #537 and~~ still need some discussion on what should be tested

adroitwhiz committed:
refactor: change minimum SVG mipmap level to 3

adroitwhiz committed:
fix: use "logical bounds" for distortion effect coords

adroitwhiz committed:
refactor: give skins a "native" & "quad" size and center

Skins' "native size" is the size that their bounds are supposed to be.
Their "quad size" is the size the quadrilateral is really rendered at.

adroitwhiz committed:
fix: calculate drawables' AABBs from "native" bounds

adroitwhiz committed:
fix: handle empty SVG skins' rotation centers

adroitwhiz committed:
fix: remove debugging bounds around rendered sprites

adroitwhiz committed:
fix: do CPU-side effect transforms in "logical space"

adroitwhiz committed:
fix: only apply CPU mosaic+pixelate in "logical bounds"

fsih commented on src/SVGSkin.js:
```diff
@@ -159,6 +159,14 @@ class SVGSkin extends Skin {
             this._rotationCenter[0] = adjustedCenter[0];
             this._rotationCenter[1] = adjustedCenter[1];
 
+            // Set "logical bounds" as fractions of the adjusted bounding box size
```
Add more commentary here

fsih commented on src/shaders/sprite.frag:
```diff
@@ -154,8 +164,16 @@ void main()
 	gl_FragColor = texture2D(u_skin, texcoord0);
 
+	#ifdef ENABLE_pixelate
+	// Ensure that the pixels don't extend outside the "logical bounds"
```
Add comment about why we aren't worried about <0

fsih commented on src/shaders/sprite.vert:
```diff
@@ -63,5 +65,6 @@ void main() {
 	#else
 	gl_Position = u_projectionMatrix * u_modelMatrix * vec4(a_position, 0, 1);
 	v_texCoord = a_texCoord;
+	v_logicalCoord = (a_texCoord - u_logicalBounds.xy) / (u_logicalBounds.zw - u_logicalBounds.xy);
```
add comment here

fsih commented on src/Skin.js:
```diff
@@ -82,26 +82,54 @@ class Skin extends EventEmitter {
+     * false if it differs.
+     * Vector-based skins like the SVGSkin and TextBubbleSkin will have different "quadrilateral" and "native" sizes.
+     */
+    get quadSizeIsNativeSize () {
```
Remove?

fsih commented on src/Drawable.js:
```diff
@@ -564,12 +568,64 @@ class Drawable {
+        }
+
+
+        result = result || this._aabb;
```
Always instantiate new instead of passing this._aabb

fsih commented on src/Drawable.js:
```diff
@@ -102,6 +102,9 @@ class Drawable {
         this._inverseTransformDirty = true;
         this._visible = true;
 
+        this._aabbDirty = true;
```
transformDirty means that one or more of rotationTransformDirty, skinScaleDirty, rotationCenterDirty.
transformDirty implies aabbDirty, but aabbDirty doesn't imply transformDirty

fsih commented on src/Drawable.js:
```diff
@@ -564,12 +568,64 @@ class Drawable {
+            const origBottom = (size[1] - center[1]) * this._scale[1] / 100;
+
+            const leftCos = origLeft * cos;
+            const rightCos = origRight * cos;
```
It bothers me that top and bottom are grouped but left and right are intermingled in the ordering

fsih commented on src/Drawable.js:
```diff
@@ -564,12 +568,64 @@ class Drawable {
+            const cos = this._rotationMatrix[0];
+            const sin = this._rotationMatrix[1];
+
+            // TODO: this can likely be golfed
```
Comment

fsih commented on src/SVGSkin.js:
```diff
@@ -156,13 +156,17 @@ class SVGSkin extends Skin {
+            return;
+        }
+
+        if (typeof rotationCenter === 'undefined') rotationCenter = this.calculateRotationCenter();
```
calculateRotationCenter () {
        return [this.size[0] / 2, this.size[1] / 2];
    }

adroitwhiz committed:
refactor: make _calculateRotationCenter private
...and just return nativeRotationCenter in getSkinRotationCenter to do
so.

fsih commented on src/EffectTransform.js:
```diff
@@ -224,8 +224,8 @@ class EffectTransform {
         }
         if ((effects & ShaderManager.EFFECT_INFO.pixelate.mask) !== 0) {
             // vec2 pixelTexelSize = u_skinSize / u_pixelate;
-            const texelX = skinUniforms.u_skinSize[0] * uniforms.u_pixelate;
```
Remember to put this separate PR in at some point

cwillisf commented on src/EffectTransform.js:
```diff
@@ -217,12 +217,15 @@ class EffectTransform {
             (skinUniforms.u_logicalBounds[3] - skinUniforms.u_logicalBounds[1]);
 
-        if ((effects & ShaderManager.EFFECT_INFO.mosaic.mask) !== 0) {
+        const pointInsideLogicalBounds = dst[0] >= 0 && dst[0] <= 1 && dst[1] >= 0 && dst[1] <= 1;
```
Would it make sense to have the caller do the remap to logical coordinates and then this check, and not even call `transformPoint` if the logical coordinate is outside the bounds? I'm worried about side effects of discontinuous transformations in (for example) convex hull calculation.

adroitwhiz replied:
Done in https://github.com/LLK/scratch-render/pull/594/commits/036895581e72045b95957e656eec64f117e8eca5

adroitwhiz committed:
refactor: use subpixel-adjusted center+size for SVG skins

adroitwhiz committed:
docs: document new rectangle AABB algorithm

adroitwhiz committed:
refactor: remove Rectangle.initFromModelMatrix

adroitwhiz committed:
refactor: reflow text bubble lines non-lazily

This was a case of premature optimization that needs to go for the next
commit's optimization to work.

adroitwhiz committed:
perf: replace skin size getters with properties

This is an optimization from PR #470 that was taken out because of other
unrelated changes in that PR which kept breaking things. We can put it
back in now, and it's especially useful here to avoid going through a
nested getter when accessing quadSize.

adroitwhiz committed:
docs: document u_logicalBounds a bit more

adroitwhiz committed:
perf: optimize "logical bounds" shader calculation

Now we only mess around with "logical bounds" in the shader if effects
which distort the sprite are enabled. This also avoids passing the extra
v_logicalCoord varying.

adroitwhiz committed:
refactor: move effect transform + bounds check out of getLocalPosition

This results in a tad bit of duplicated code, but considering that we
have 4 different code paths (colorAtNearest, colorAtLinear,
isTouchingNearest, and isTouchingLinear) for sampling a silhouette,
that's to be expected.

This more closely matches the GPU pipeline, in which color and position
calculations are more intertwined. This replaces the hacky fix in #424
with a solution that matches the GPU: instead of not transforming points
outside the skin bounds, just return transparency/false early.

adroitwhiz committed:
perf: use _skin instead of getter in "touching" methods

This provides a slight but noticeable performance improvement, and gains
back some of what was lost in the previous commit.

adroitwhiz committed:
refactor: remove createMIP zero-size checks

_setSubpixelViewbox ensures nonzero image dimensions at all rendered
scales.

adroitwhiz commented:
@fsih @cwillisf This should be ready now that scratch-svg-renderer is all split up and we can do the viewbox logic here! I've also tried my best to address the review comments. Also note that this depends on #745.

I remember you not liking the names `quadSize` and `quadRotationCenter`--any ideas what I could use instead? `textureSize`/`textureRotationCenter` seems misleading because that seems to refer to the texture's actual dimensions, which isn't what `quadSize` and `rotationCenter` refer to.

adroitwhiz committed:
docs: document subtlety in setting SVG skin dimensions
